iina/Logger.swift

```diff
@@ -0,0 +1,148 @@
+//
+//  Logger.swift
+//  iina
+//
+//  Created by Collider LI on 24/5/2018.
+//  Copyright © 2018 lhc. All rights reserved.
+//
+
+import Foundation
+
+class Logger {
+
+  enum LogLevel: Int, Comparable {
+
+    static func < (lhs: LogLevel, rhs: LogLevel) -> Bool {
+      return lhs.rawValue < rhs.rawValue
+    }
+
+    case verbose = 0
+    case debug
+    case warning
+    case error
+
+    var stringValue: String {
+      switch self {
+      case .verbose: return "v"
+      case .debug: return "d"
+      case .warning: return "w"
+      case .error: return "e"
+      }
+    }
+  }
+
+  var label: String
+  var level: LogLevel
+  var useNSLog = false
+
+  static let enabled = Preference.bool(for: .enableLogging)
+  static let logDirectory: URL = {
+    let date = Date()
+    let calendar = NSCalendar.current
+    let y = calendar.component(.year, from: date)
+    let m = calendar.component(.month, from: date)
+    let d = calendar.component(.day, from: date)
+    let h = calendar.component(.hour, from: date)
+    let mm = calendar.component(.minute, from: date)
+    let s = calendar.component(.second, from: date)
+    let token = Utility.ShortCodeGenerator.getCode(length: 6)
+    let sessionDirName = "\(y)-\(m)-\(d)-\(h)-\(mm)-\(s)_\(token)"
+    let sessionDir = Utility.logDirURL.appendingPathComponent(sessionDirName, isDirectory: true)
+    Utility.createDirIfNotExist(url: sessionDir)
+    return sessionDir
+  }()
+
+  private static var logFileHandle: FileHandle = {
+    let logFileURL = logDirectory.appendingPathComponent("iina.log")
+    FileManager.default.createFile(atPath: logFileURL.path, contents: nil, attributes: nil)
+    return try! FileHandle(forWritingTo: logFileURL)
+  }()
+
+  private static var dateFormatter: DateFormatter = {
+    let formatter = DateFormatter()
+    formatter.dateFormat = "dd HH:mm:ss.SSS"
+    return formatter
+  }()
+
+  static func getLogger(_ label: String) -> Logger? {
```

Comment:
In the current implementation, we may have multiple logger for different components.

Reply:
How does this differ from passing in a different label?

Reply:
We can initialize a logger for certain part and use it instead of pass in same label over and over again.

Reply:
Alas, we can't curry functions in Swift anymore, or this might have a nicer solution. Ok: we can have `Logger` be intializable.

Reply:
- We don't need to type the same label again and again and be careful of typos when using separate loggers.
- This is more flexible in cases like having multiple `PlayerCore`s, so that each `PlayerCore` can have its own logger with a dynamic label based on its ID or index.
- Different `Logger`s can have its own configuration, like log level, whether log to console, etc.